ddtrace/contrib/langchain/patch.py

```diff
@@ -0,0 +1,752 @@
+            return
+        for token_type in ("prompt", "completion", "total"):
+            num_tokens = usage.get("token_usage", {}).get(token_type + "_tokens")
+            if not num_tokens:
+                continue
+            self.metric(span, "dist", "tokens.%s" % token_type, num_tokens)
+        total_cost = span.get_metric("langchain.tokens.total_cost")
+        if total_cost:
+            self.metric(span, "incr", "tokens.total_cost", total_cost)
+
+
+def _extract_model_name(instance):
+    # type: (langchain.llm.BaseLLM) -> Optional[str]
+    """Extract model name or ID from llm instance."""
+    for attr in ("model", "model_name", "model_id", "model_key", "repo_id"):
+        if hasattr(instance, attr):
+            return getattr(instance, attr)
+    return None
+
+
+def _format_api_key(api_key):
+    # type: (str) -> str
+    """Obfuscate a given LLM provider API key by returning the last four characters."""
+    if not api_key or len(api_key) < 4:
+        return ""
+    return "...%s" % api_key[-4:]
+
+
+def _extract_api_key(instance):
+    # type: (Any) -> str
+    """
+    Extract and format LLM-provider API key from instance.
+    Note that langchain's LLM/ChatModel/Embeddings interfaces do not have a
+    standard attribute name for storing the provider-specific API key, so make a
+    best effort here by checking for attributes that end with `api_key/api_token`.
+    """
+    api_key_attrs = [a for a in dir(instance) if a.endswith(("api_token", "api_key"))]
+    if api_key_attrs and hasattr(instance, str(api_key_attrs[0])):
+        api_key = getattr(instance, api_key_attrs[0], None)
+        if api_key:
+            return _format_api_key(api_key)
+    return ""
+
+
+def _tag_openai_token_usage(span, llm_output, propagated_cost=0, propagate=False):
+    # type: (Span, Dict[str, Any], int, bool) -> None
+    """
+    Extract token usage from llm_output, tag on span.
+    Calculate the total cost for each LLM/chat_model, then propagate those values up the trace so that
+    the root span will store the total token_usage/cost of all of its descendants.
+    """
+    for token_type in ("prompt", "completion", "total"):
+        current_metric_value = span.get_metric("langchain.tokens.%s_tokens" % token_type) or 0
+        metric_value = llm_output["token_usage"].get("%s_tokens" % token_type, 0)
+        span.set_metric("langchain.tokens.%s_tokens" % token_type, current_metric_value + metric_value)
+    total_cost = span.get_metric("langchain.tokens.total_cost") or 0
+    if not propagate:
+        try:
+            completion_cost = get_openai_token_cost_for_model(
+                span.get_tag("langchain.request.model"),
+                span.get_metric("langchain.tokens.completion_tokens"),
+                is_completion=True,
+            )
+            prompt_cost = get_openai_token_cost_for_model(
+                span.get_tag("langchain.request.model"), span.get_metric("langchain.tokens.prompt_tokens")
```

Comment:
We can use the MODEL constant here (this could be overkill but we can add a constant for token too)